core-primitives/stf-state-observer/src/state_observer.rs

```diff
@@ -0,0 +1,153 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	traits::{ObserveState, UpdateState},
+};
+use itp_types::ShardIdentifier;
+use std::{collections::HashMap, vec::Vec};
+
+/// State observer implementation. Receives updates in a dedicated queue.
+/// These updates are applied every time an observation function is executed.
+///
+#[derive(Default)]
+pub struct StateObserver<StateType> {
+	queued_state_updates: RwLock<HashMap<ShardIdentifier, StateType>>,
+	current_state: RwLock<HashMap<ShardIdentifier, StateType>>,
+}
+
+impl<StateType> StateObserver<StateType> {
+	pub fn new(maybe_state: Option<(ShardIdentifier, StateType)>) -> Self {
+		let initial_hash_map = match maybe_state {
+			None => HashMap::new(),
+			Some((shard, state)) => HashMap::from([(shard, state)]),
+		};
+
+		Self {
+			queued_state_updates: Default::default(),
+			current_state: RwLock::new(initial_hash_map),
+		}
+	}
```

Comment:
Hmm, some nitpicking here, wouldn't it be better to split this into two methods and do:

1. derive default
2. change the new signature to: `pub fn new(shard: ShardIdentifier, state: StateType) -> Self`

Reply:
It already derives `Default`. But yes, changing the signature is probably an improvement 👍